src/Microsoft.IdentityModel.Tokens/RsaSecurityKey.cs

```diff
@@ -27,69 +27,110 @@
 
 using System;
 using System.Security.Cryptography;
+using Microsoft.IdentityModel.Logging;
 
 namespace Microsoft.IdentityModel.Tokens
 {
     public class RsaSecurityKey : AsymmetricSecurityKey
     {
-        private RSAParameters _rsaParamaeters;
+        PrivateKeyStatus _privateKeyStatus = PrivateKeyStatus.AvailabilityNotDetermined;
 
         public RsaSecurityKey(RSAParameters rsaParameters)
         {
-            // must have private or public key
-            bool hasPrivateKey = rsaParameters.D != null && rsaParameters.DP != null && rsaParameters.DQ != null && rsaParameters.P != null && rsaParameters.Q != null;
-            bool hasPublicKey = rsaParameters.Exponent != null && rsaParameters.Modulus != null;
-            if (!hasPrivateKey && !hasPublicKey)
-            {
-                throw new ArgumentException("no public or private key material found");
-            }
+            // must have modulus and exponent otherwise the crypto operations fail later
+            if (rsaParameters.Modulus == null || rsaParameters.Exponent == null)
+                throw LogHelper.LogException<ArgumentException>(LogMessages.IDX10700, rsaParameters.ToString());
 
-            _rsaParamaeters = rsaParameters;
+            _privateKeyStatus = rsaParameters.D != null && rsaParameters.DP != null && rsaParameters.DQ != null && rsaParameters.P != null && rsaParameters.Q != null && rsaParameters.InverseQ != null ? PrivateKeyStatus.HasPrivateKey : PrivateKeyStatus.DoesNotHavePrivateKey;
+            Parameters = rsaParameters;
         }
 
-        public override int KeySize
+        public RsaSecurityKey(RSA rsa)
+        {
+            if (rsa == null)
+                throw LogHelper.LogArgumentNullException("rsa");
+
+            Rsa = rsa;
+            if (HasPrivateKey)
+                Parameters = Rsa.ExportParameters(true);
+            else
+                Parameters = Rsa.ExportParameters(false);
+        }
+
+        public override bool HasPrivateKey
```

Comment:
I am going to have another PR coming soon (after we discuss the possible solutions on the respective issues: #337 #330 #336). Will handle this issue then.